openhcl/hcl/src/ioctl.rs

```diff
@@ -1835,29 +1836,50 @@ impl<'a, T: Backing> ProcessorRunner<'a, T> {
     /// Gets the proxied interrupt request bitmap from the hypervisor.
     pub fn proxy_irr(&mut self) -> Option<[u32; 8]> {
         // SAFETY: the `scan_proxy_irr` and `proxy_irr` fields of the run page
-        // are concurrently updated by the kernel, but only on this processor.
+        // are concurrently updated by the kernel on multiple processors. They
+        // are accessed atomically everywhere.
         unsafe {
             let scan_proxy_irr =
                 &*(addr_of!((*self.run.as_ptr()).scan_proxy_irr).cast::<AtomicU8>());
             let proxy_irr = &*(addr_of!((*self.run.as_ptr()).proxy_irr).cast::<[AtomicU32; 8]>());
-            if scan_proxy_irr.load(Ordering::Relaxed) == 0 {
+            if scan_proxy_irr.load(Ordering::Acquire) == 0 {
                 return None;
             }
 
-            scan_proxy_irr.store(0, Ordering::Relaxed);
+            scan_proxy_irr.store(0, Ordering::Release);
             let mut r = [0; 8];
             for (irr, r) in proxy_irr.iter().zip(r.iter_mut()) {
-                // In theory we don't need atomic operations, we just need the
-                // swap to happen in a single instruction. This would require
-                // inline assembly.
-                if irr.load(Ordering::Relaxed) != 0 {
-                    *r = irr.swap(0, Ordering::Relaxed);
+                if irr.load(Ordering::Acquire) != 0 {
+                    *r = irr.swap(0, Ordering::SeqCst);
```

Comment:
I think you can keep `Relaxed` here since this is all guaranteed to happen after the `scan_proxy_irr` store above (once you upgrade that to SeqCst).

Reply:
Fixed